File: shortestpath3.cpp
#include <iostream>
#include <vector>
using namespace std;

const int inf = 1 << 30;

int main() {

    ios_base::sync_with_stdio(false); cin.tie(0); cout.tie(0);

    int n,m,q,s;
    while(1){
        cin >>n>>m>>q>>s;
        if(!n&&!m&&!s&&!q) break;
        vector< vector<int> > edges;
        vector<int> dis(n, inf);
        dis[s]=0;
        int u,v,w,i;
        for(i = 0; i < m; ++i) {
            cin>>u>>v>>w;
            edges.push_back({u,v,w});
        }

        // bellman-ford
        for(i=1; i<n; ++i) {
            for(auto& v : edges) {
                if (dis[v[0]]!=inf && dis[v[0]]+v[2]<dis[v[1]]) {
                    dis[v[1]]=dis[v[0]]+v[2];
                }
            }
        }
        // for(int e : dis) cout << e <<" "; cout << endl;

        // negative cycle
        bool cycle=false;
        for(auto& v : edges) {
            if(dis[v[0]]==-inf || (dis[v[0]]!=inf && dis[v[0]]+v[2]<dis[v[1]])) {
                dis[v[1]]=-inf;
                cycle=true;
            }
        }

        if(cycle){
            for(i=2;i<n;++i){
                for(auto& v : edges) {
                    if(dis[v[0]]==-inf || (dis[v[0]]!=inf && dis[v[0]]+v[2]<dis[v[1]])) 
                        dis[v[1]]=-inf;
                } 
            }
        }

        int dest;
        for(i=0;i<q;++i) {
            cin>>dest;
            if(dis[dest]==-inf) cout<<"-Infinity"<<endl;
            else if(dis[dest]==inf) cout<<"Impossible"<<endl;
            else cout << dis[dest]<<endl;
        }
        cout << endl;
    }
}

File: ShortestPath3.java
import java.util.*;

public class ShortestPath3 {

    static class Pair {
        int u;
        int v;
        int weight;
        public Pair(int u, int v, int weight) {
            this.u = u;
            this.v = v;
            this.weight = weight;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (true) {
            int n = sc.nextInt();
            int m = sc.nextInt();
            int q = sc.nextInt();
            int s = sc.nextInt();

            if (n == 0 && m == 0 && q == 0 && s == 0) break;

            int[] dis = new int[n];
            List<Pair> list = new LinkedList<>();
            Arrays.fill(dis, Integer.MAX_VALUE);
            dis[s] = 0;
            for (int i = 0; i < m; ++i) {
                int u = sc.nextInt();
                int v = sc.nextInt();
                int w = sc.nextInt();

                list.add(new Pair(u, v, w));
            }

            // relax
            for (int i = 1; i < n; ++i) {
                for (Pair p : list) {
                    if (dis[p.u] != Integer.MAX_VALUE && dis[p.u] + p.weight < dis[p.v]) {
                        dis[p.v] = dis[p.u] + p.weight;
                    }
                }
            }

            // detect negative cycle
            boolean negative = false;
            for (Pair p : list) {
                if (dis[p.u] == Integer.MIN_VALUE || (dis[p.u] != Integer.MAX_VALUE && dis[p.u] + p.weight < dis[p.v])) {
                    dis[p.v] = Integer.MIN_VALUE;
                    negative = true;
                }
            }

            if (negative) {
                for (int i = 1; i < n; ++i) {
                    for (Pair p : list) {
                        if (dis[p.u] == Integer.MIN_VALUE || (dis[p.u] != Integer.MAX_VALUE && dis[p.u] + p.weight < dis[p.v])) {
                            dis[p.v] = Integer.MIN_VALUE;
                        }
                    }
                }
            }

            for (int i = 0; i < q; ++i) {
                int next = sc.nextInt();
                if (dis[next] == Integer.MIN_VALUE) {
                    System.out.println("-Infinity");
                } else if (dis[next] == Integer.MAX_VALUE) {
                    System.out.println("Impossible");
                } else {
                    System.out.println(dis[next]);
                }
            }
            System.out.print("\n");
        }
    }
}
